lib/src/builder/parsing/version.dart

```diff
@@ -0,0 +1,240 @@
+///
+/// As the builder detects members, it generates a confidence based on the
+/// implementation details of the member. This confidence score is then averaged
+/// over all of the grouped members (factory, props, state, component, etc).
+///
+/// Related: [resolveVersion]
+@sealed
+abstract class Confidence {
+  /// Not possible to be considered a given version.
+  static const none = 0;
+
+  /// Unlikely to be considered a given version.
+  ///
+  /// This value is an order of magnitude less than [neutral].
+  static const unlikely = 1;
+
+  /// Equally likely and unlikely to be considered a given version.
+  ///
+  /// Consuming code should only consider values strictly greater than this value
+  /// as confident enough to perform code generation:
+  ///
+  ///    - `confidence <= neutral`: do not perform code generation
+  ///    - `confidence > neutral`: perform code generation
+  ///
+  /// This value is an order of magnitude both greater than [unlikely] and less than [likely].
+  static const neutral = 10;
+
+  /// Likely to be considered a given version.
+  ///
+  /// This value is an order of magnitude both greater than [neutral] and less than [certain].
+  static const likely = 100;
+
+  /// Certain to be considered a given version.
+  ///
+  /// This value is an order of magnitude greater than [likely].
+  static const certain = 1000;
+
+  static const _confidenceToString = {
+    none: 'none',
+    unlikely: 'unlikely',
+    neutral: 'neutral',
+    likely: 'likely',
+    certain: 'certain',
+  };
+
+  /// Returns a string description of a [confidence] score, relating it to the constants in this
+  /// class.
+  ///
+  /// Example:
+  ///
+  ///     - Confidence.description(Confidence.neutral)    // 'neutral'
+  ///     - Confidence.description(Confidence.likely)     // 'likely'
+  ///     - Confidence.description(Confidence.likely + 1) // 'likely + 1'
+  static String description(int confidence) {
+    final baseConfidence = _confidenceToString.keys.where((c) => confidence >= c).reduce(max);
+    final difference = confidence - baseConfidence;
+
+    return '${_confidenceToString[baseConfidence]}${difference == 0 ? '' : ' + $difference'}';
+  }
+}
+
+/// The current versions of the boilerplate that the builder detects.
+///
+/// Related: [resolveVersion]
+enum Version {
+  // "v1" is for the transformer-based syntax :)
+  v2_legacyBackwardsCompat,
+  v3_legacyDart2Only,
+  v4_mixinBased,
+}
+
+/// Priority of the possible boilerplate versions.
+///
+/// In the event that two different [Version]s have the same confidence score,
+/// version of high priority wins out.
+///
+/// The length of this iterable should match the length of values in [Version].
+const _versionsInPriorityOrder = [
+  Version.v4_mixinBased,
+  Version.v2_legacyBackwardsCompat,
+  Version.v3_legacyDart2Only,
+];
+
+/// Utils to add capabilities to the [Version] enum.
+extension VersionExtension on Version {
```

Comment:
No way... I didn't know you could do `enum` extensions!!!??!?! Sick!

Reply:
Right???